omp/preconditioner/batch_jacobi_kernels.cpp

```diff
@@ -0,0 +1,201 @@
+        const auto z_b = gko::batch::extract_batch_item(zub, batch_id);
+
+        const auto work_arr_size = PrecType::dynamic_work_size(
+            sys_mat_batch.num_rows, sys_mat_batch.get_single_item_num_nnz());
+        std::vector<ValueType> work(work_arr_size);
+
+        prec.generate(batch_id, sys_mat_entry, work.data());
+        prec.apply(r_b, z_b);
+    }
+}
+
+
+// Note: Do not change the ordering
+#include "reference/preconditioner/batch_jacobi_kernels.hpp.inc"
+
+
+}  // unnamed namespace
+
+
+template <typename ValueType, typename IndexType>
+void batch_jacobi_apply(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const batch::matrix::Csr<ValueType, IndexType>* const sys_mat,
+    const size_type num_blocks, const uint32 max_block_size,
+    const gko::batch::preconditioner::batched_jacobi_blocks_storage_scheme<
+        IndexType>& storage_scheme,
+    const IndexType* const cumulative_block_storage,
+    const ValueType* const blocks_array, const IndexType* const block_ptrs,
+    const IndexType* const row_part_of_which_block_info,
+    const batch::MultiVector<ValueType>* const r,
+    batch::MultiVector<ValueType>* const z)
+{
+    const auto sys_mat_batch = host::get_batch_struct(sys_mat);
+    batch_jacobi_apply_helper(sys_mat_batch, num_blocks, max_block_size,
+                              storage_scheme, cumulative_block_storage,
+                              blocks_array, block_ptrs,
+                              row_part_of_which_block_info, r, z);
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INT32_TYPE(
+    GKO_DECLARE_BATCH_JACOBI_APPLY_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void batch_jacobi_apply(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const batch::matrix::Ell<ValueType, IndexType>* const sys_mat,
+    const size_type num_blocks, const uint32 max_block_size,
+    const gko::batch::preconditioner::batched_jacobi_blocks_storage_scheme<
+        IndexType>& storage_scheme,
+    const IndexType* const cumulative_block_storage,
+    const ValueType* const blocks_array, const IndexType* const block_ptrs,
+    const IndexType* const row_part_of_which_block_info,
+    const batch::MultiVector<ValueType>* const r,
+    batch::MultiVector<ValueType>* const z)
+{
+    const auto sys_mat_batch = host::get_batch_struct(sys_mat);
+    batch_jacobi_apply_helper(sys_mat_batch, num_blocks, max_block_size,
+                              storage_scheme, cumulative_block_storage,
+                              blocks_array, block_ptrs,
+                              row_part_of_which_block_info, r, z);
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INT32_TYPE(
+    GKO_DECLARE_BATCH_JACOBI_ELL_APPLY_KERNEL);
+
+
+template <typename IndexType>
+void compute_cumulative_block_storage(
+    std::shared_ptr<const DefaultExecutor> exec, const size_type num_blocks,
+    const IndexType* const block_pointers,
+    IndexType* const blocks_cumulative_storage)
+{
+#pragma omp parallel for
```

Comment:
I would assume that `num_blocks << 1000`, since the matrices are usually small, so maybe using the parallel for would be even slower than no parallelization.

Reply:
Matrix sizes can go upto 2000. For small block sizes, ~2, the number of blocks can be around 2000. In any case, I dont think the parallelization should be a problem for the small cases.